lib/SILGen/SILGenType.cpp

```diff
@@ -109,13 +109,21 @@ SILGenModule::emitVTableMethod(ClassDecl *theClass,
   
   auto overrideInfo = M.Types.getConstantOverrideInfo(derived, base);
 
+  // If base method's generic requirements are not satisfied by the derived
+  // method then we need a thunk.
+  auto sig = getASTContext().getOverrideGenericSignature(base.getDecl(),
+                                                         derived.getDecl());
+  auto derivedSig =
+      derived.getDecl()->getAsGenericContext()->getGenericSignature();
+  auto needsThunk = sig && !sig->requirementsNotSatisfiedBy(derivedSig).empty();
```

Comment:
requirementsNotSatisfiedBy().empty() is called in a few places by now. Do you mind factoring it out into a new utility method?

Or if you want to be really fancy -- the whole getOverrideGenericSignature() / requirementsNotSatisfiedBy() dance could be a utility method somewhere.

Also IMO needsThunk could have a more descriptive name. Right now it's confusing that needsThunk can be false even when we need a thunk for other reasons. Perhaps 'bool hasGenericRequirementDifference'?